File: WhatsAppAutoReply/app/src/main/java/com/whatsappauto/reply/database/ScheduledMessageDao.java
package com.whatsappauto.reply.database;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import com.whatsappauto.reply.models.ScheduledMessage;

import java.util.List;

@Dao
public interface ScheduledMessageDao {
    @Insert
    void insert(ScheduledMessage scheduledMessage);

    @Update
    void update(ScheduledMessage scheduledMessage);

    @Delete
    void delete(ScheduledMessage scheduledMessage);

    @Query("SELECT * FROM scheduled_messages WHERE isSent = 0 ORDER BY scheduledTime ASC")
    LiveData<List<ScheduledMessage>> getPendingMessages();

    @Query("SELECT * FROM scheduled_messages ORDER BY scheduledTime DESC")
    LiveData<List<ScheduledMessage>> getAllMessages();

    @Query("SELECT * FROM scheduled_messages WHERE scheduledTime <= :currentTime AND isSent = 0")
    List<ScheduledMessage> getMessagesToSend(long currentTime);

    @Query("UPDATE scheduled_messages SET isSent = 1 WHERE id = :id")
    void markAsSent(int id);
}

File: app/src/main/java/com/whatsappautoresponder/database/ScheduledMessageDao.kt
package com.whatsappautoresponder.database

import androidx.lifecycle.LiveData
import androidx.room.*

@Dao
interface ScheduledMessageDao {
    
    @Query("SELECT * FROM scheduled_messages ORDER BY scheduledTime ASC")
    fun getAllScheduledMessages(): LiveData<List<ScheduledMessage>>
    
    @Query("SELECT * FROM scheduled_messages WHERE isCompleted = 0 AND scheduledTime <= :currentTime ORDER BY scheduledTime ASC")
    suspend fun getPendingMessages(currentTime: Long): List<ScheduledMessage>
    
    @Query("SELECT * FROM scheduled_messages WHERE id = :id")
    suspend fun getScheduledMessageById(id: Long): ScheduledMessage?
    
    @Insert
    suspend fun insertScheduledMessage(message: ScheduledMessage): Long
    
    @Update
    suspend fun updateScheduledMessage(message: ScheduledMessage)
    
    @Delete
    suspend fun deleteScheduledMessage(message: ScheduledMessage)
    
    @Query("DELETE FROM scheduled_messages WHERE id = :id")
    suspend fun deleteScheduledMessageById(id: Long)
    
    @Query("UPDATE scheduled_messages SET isCompleted = 1 WHERE id = :id")
    suspend fun markMessageCompleted(id: Long)
    
    @Query("SELECT * FROM scheduled_messages WHERE isRecurring = 1 AND isCompleted = 1")
    suspend fun getCompletedRecurringMessages(): List<ScheduledMessage>
}